libass/ass_directwrite.c

```diff
@@ -254,24 +254,41 @@ static void init_FallbackLogTextRenderer(FallbackLogTextRenderer *r,
 }
 
 /*
- * This function is called whenever a font is accessed for the
- * first time. It will create a FontFace for metadata access and
- * memory reading, which will be stored within the private data.
+ * A helper function for init_font_private_stream and get_utf8_path.
+ * Remember to release the loader after obtaining it.
  */
-static bool init_font_private_face(FontPrivate *priv)
+static bool get_font_file_loader_and_reference_key(
+    FontPrivate *priv,
+    const void **refKey,
+    UINT32* keySize,
+    IDWriteFontFileLoader **loader
+    )
 {
-    HRESULT hr;
-    IDWriteFontFace *face;
+    bool ret = false;
+    HRESULT hr = S_OK;
+    IDWriteFontFile *file = NULL;
+    UINT32 n_files = 1;
 
-    if (priv->face != NULL)
-        return true;
+    /* DirectWrite only supports one file per face */
+    hr = IDWriteFontFace_GetFiles(priv->face, &n_files, &file);
+    if (FAILED(hr) || !file)
+        goto cleanup;
 
-    hr = IDWriteFont_CreateFontFace(priv->font, &face);
-    if (FAILED(hr) || !face)
-        return false;
+    hr = IDWriteFontFile_GetReferenceKey(file, refKey, keySize);
+    if (FAILED(hr))
+        goto cleanup;
 
-    priv->face = face;
-    return true;
+    hr = IDWriteFontFile_GetLoader(file, loader);
+    if (FAILED(hr) || !*loader)
+        goto cleanup;
+
+    ret = true;
+
+cleanup:
+    if (file)
+        IDWriteFontFile_Release(file);
```

Comment:
Ah, I’m afraid this isn’t as easy as it seemed.

You can’t release the file here on the success path, because this potentially invalidates the reference key that you’re returning: its lifetime is tied to the file’s lifetime. So then you have to move this release back into the callers, which means you must return the file itself via another output argument, which reduces the gains from factoring this out in the first place. But at least the three API calls above are shared, so perhaps it’s still beneficial?

I also wanted to suggest returning the loader directly instead of a Boolean, but if you need to return both the loader and the file anyway, then the Boolean might be cleaner.

Reply:
Hmm fair. The fact that it works probably means that IDWriteFontFace keeps references to file alive, but we probably shouldn't rely on that at all.

I think the boolean is cleaner because it saves us from potentially checking values of the other params if it returns true (otherwise someone would look at the invocation and think the coder forgot to check them if it returned just the object instead)

Reply:
This leaves the question on what to do in case of failure in terms of cleanup. I think I will just leave the cleanup to the callers to avoid code duplication there...

Reply:
You _might_ want to explicitly initialize `*file` to `NULL` inside the function for extra clarity/safety, although, of course, we do currently null-initialize it in the callers anyway.

Reply:
(This is turning into bikeshedding, and the code is good for merge as it is, but:)

Hm, I don’t think it’s useful to initialize the key reference variables here, nor for that matter the loader, because if the function fails, those aren’t supposed to be used by the caller in any way. But the file specifically, by your contract, must be freed by the caller even if this function fails—so my logic was that it should always have a valid value even if this function fails.

Reply:
> But the file specifically, by your contract, must be freed by the caller even if this function fails

Well in case of failure it's only file, but in case of success both file and loader. In the parent funcs we implement it by simply null-checking for both and then freeing them, and in this situation it's probably better to NULL-initialize them both (otherwise we risk someone using the func in some 3rd place down the line, without NULL-setting the input vars, and they'd have a bug unbeknownst to them where they'd if-check for non-NULL-initialized argument and try to Release it, if they copy-pasted the `cleanup` section).

And then when we set 2 of 4 params to null, we may as well just set all 4 for consistency xD I think out of nulling 0/4 or 4/4, nulling only 2/4 would be the least intuitive

Reply:
Uhh, not quite sure I follow: upon success, the loader _is_ filled—regardless of the added null-initialization.

Reply:
Yeah, but on early failure it could remain untouched without being set to null in helper func.

Then if some kind of hypothetical new parent func passed arg of `IDWriteFontFileLoader * loader;` without NULL-initializing it itself, and then did `if (loader) IDWriteFontFileLoader_Release(loader);`, then they could potentially be trying to Release uninitialized garbage and get a crash (but they'd miss the bug, because normally the func suceeds).

And with that I believe it's beneficial to also null it in the helper as well, just like you said to do to file, if we're doing that already now anyways

Reply:
To me, there’s a clear difference in the API contract:

If the function call fails, it doesn’t produce a `loader`, so it makes no promises about its value, so the caller mustn’t just `if (loader)` unless the caller itself has zero-init it. Just like in our calls to DirectWrite.

Whereas for `file`, this function _requires_ the caller to handle it both on success and on failure, so it is this function’s own duty to ensure that it actually is able to be handled by the caller in every case. It returns a `loader` only upon success, but it returns a `file` _always._

Of course, it can also promise that it always fills the `loader`, as your code does now, and that’s a valid choice—just not one I would personally take in this context. But like I said, this doesn’t even matter all that much.